### ISSUE TITLE
Deadlock in CurlHandleContainer::AcquireCurlHandle()

### ISSUE THREAD

**Issue by amitojsinghbhatia** [aws/aws-sdk-cpp#1456]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/welcome.html) and [API reference](http://sdk.amazonaws.com/cpp/api/LATEST/index.html)
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-cpp/issues) and didn't find any solution
  
**Describe the bug**
When a heavily multi-threaded application creates a lot of S3Clients, a deadlock is observed in CurlHandleContainer::AcquireCurlHandle()
This seems related to #713 but that is an old bug report, and it was deemed there would be only be sub-optimal pool growth and no hang, while I am actually observing a hang, so opened a new issue.

**SDK version number**
1.8.0

**Platform/OS/Hardware/Device**
Linux x86_64 - CentOS 6.10

**To Reproduce (observed behavior)**
Create 50 S3Clients in separate threads, or just run `httpClient->MakeRequest()` in parallel. I am not sure if the number 50 is specific to my platform or is a general limit as per pool max size.
I created a PR #1453 with a fix (Sorry, I read the guidelines and should have opened the bug first and submitted PR later). Anyway, the PR has a unit testcase which reproduces the issue.

**Expected behavior**
No hang.

**Additional context**
Here is what I think is the cause. Consider this snippet from AcquireCurlHandle:

    if(!m_handleContainer.HasResourcesAvailable())   
    {  
        CheckAndGrowPool();  
    }  
    CURL* handle = m_handleContainer.Acquire();

If the pool has already reached max limit, it will not grow further and caller will block at `m_handleContainer.Acquire();`
However, if the thread that has acquired the handle, encounters an error, it calls DestroyCurlHandle() instead of ReleaseCurlHandle(), which does not release the handle back into the pool. If none of the handles are released back, then the waiting threads will never get the handle.

**Comment by KaibaLopez** [comment]
Hi @amitojsinghbhatia ,
Thanks for submitting this, we'll take a look and review the pr, it all sounds reasonable and really appreciate the test case too, makes our life much easier.

**Comment by KaibaLopez** [comment]
Fix PR has been merged.
